// bk-magic-vue/lib/navigation.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('bk-magic-vue/lib/locale')) :
    typeof define === 'function' && define.amd ? define(['exports', 'bk-magic-vue/lib/locale'], factory) :
    (global = global || self, factory(global.library = {}, global.locale));
}(this, function (exports, locale) { 'use strict';

    var script = {
      name: 'navigation-menu-title',
      props: {
        sideTitle: {
          type: String,
          default: ''
        },
        navigationType: {
          type: String,
          default: 'left-right'
        }
      }
    };

    function normalizeComponent(template, style, script, scopeId, isFunctionalTemplate, moduleIdentifier
    , shadowMode, createInjector, createInjectorSSR, createInjectorShadow) {
      if (typeof shadowMode !== 'boolean') {
        createInjectorSSR = createInjector;
        createInjector = shadowMode;
        shadowMode = false;
      }
      var options = typeof script === 'function' ? script.options : script;
      if (template && template.render) {
        options.render = template.render;
        options.staticRenderFns = template.staticRenderFns;
        options._compiled = true;
        if (isFunctionalTemplate) {
          options.functional = true;
        }
      }
      if (scopeId) {
        options._scopeId = scopeId;
      }
      var hook;
      if (moduleIdentifier) {
        hook = function hook(context) {
          context = context ||
          this.$vnode && this.$vnode.ssrContext ||
          this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext;
          if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
            context = __VUE_SSR_CONTEXT__;
          }
          if (style) {
            style.call(this, createInjectorSSR(context));
          }
          if (context && context._registeredComponents) {
            context._registeredComponents.add(moduleIdentifier);
          }
        };
        options._ssrRegister = hook;
      } else if (style) {
        hook = shadowMode ? function () {
          style.call(this, createInjectorShadow(this.$root.$options.shadowRoot));
        } : function (context) {
          style.call(this, createInjector(context));
        };
      }
      if (hook) {
        if (options.functional) {
          var originalRender = options.render;
          options.render = function renderWithStyleInjection(h, context) {
            hook.call(context);
            return originalRender(h, context);
          };
        } else {
          var existing = options.beforeCreate;
          options.beforeCreate = existing ? [].concat(existing, hook) : [hook];
        }
      }
      return script;
    }
    var normalizeComponent_1 = normalizeComponent;

    /* script */
    var __vue_script__ = script;
    /* template */

    var __vue_render__ = function __vue_render__() {
      var _vm = this;

      var _h = _vm.$createElement;

      var _c = _vm._self._c || _h;

      return _c('div', _vm._b({
        staticClass: "bk-navigation-title",
        style: {
          borderBottomWidth: _vm.navigationType === 'left-right' ? '0' : '1px'
        }
      }, 'div', _vm.$attrs, false), [_vm._t("default", [_c('span', {
        staticClass: "title-icon"
      }, [_vm._t("side-icon", [_c('i', {
        staticClass: "bk-icon icon-rtx"
      })])], 2), _c('span', {
        staticClass: "title-desc"
      }, [_vm._v(_vm._s(_vm.sideTitle))])])], 2);
    };

    var __vue_staticRenderFns__ = [];
    /* style */

    var __vue_inject_styles__ = undefined;
    /* scoped */

    var __vue_scope_id__ = undefined;
    /* module identifier */

    var __vue_module_identifier__ = undefined;
    /* functional template */

    var __vue_is_functional_template__ = false;
    /* style inject */

    /* style inject SSR */

    /* style inject shadow dom */

    var __vue_component__ = /*#__PURE__*/normalizeComponent_1({
      render: __vue_render__,
      staticRenderFns: __vue_staticRenderFns__
    }, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, undefined, undefined, undefined);

    var script$1 = {
      name: 'bk-navigation',
      components: {
        NavigationMenuTitle: __vue_component__
      },
      props: {
        navWidth: {
          type: [Number, String],
          default: 60
        },
        hoverWidth: {
          type: [Number, String],
          default: 260
        },
        sideTitle: {
          type: String,
          default: ''
        },
        headerTitle: {
          type: String,
          default: ''
        },
        hoverLeaveDelay: {
          type: [Number, String],
          default: 0
        },
        hoverEnterDelay: {
          type: [Number, String],
          default: 100
        },
        defaultOpen: Boolean,
        headThemeColor: {
          type: String,
          default: '#182132'
        },
        themeColor: {
          type: String,
          default: '#182132'
        },
        headHeight: {
          type: [Number, String],
          default: 52
        },
        navigationType: {
          type: String,
          default: 'left-right',
          validator: function validator(v) {
            return ['top-bottom', 'left-right'].includes(v);
          }
        },
        needMenu: {
          type: Boolean,
          default: true
        }
      },
      provide: function provide() {
        return {
          rootNavigation: this
        };
      },
      data: function data() {
        return {
          defaultHeaderTitle: '',
          nav: {
            click: false,
            hover: false,
            delay: false,
            timer: null,
            enterTimer: null
          }
        };
      },
      computed: {
        defaultThemeColor: {
          get: function get() {
            var defTheme = ['#182132', '#FFFFFF'];
            return !defTheme.includes(this.themeColor) ? this.themeColor : defTheme[+(this.navigationType === 'top-bottom')];
          }
        }
      },
      created: function created() {
        this.defaultHeaderTitle = this.headerTitle || locale.t('bk.navigation.headerTitle');
        if (this.defaultOpen) {
          this.nav.click = !this.nav.click;
          this.nav.hover = this.nav.click;
          this.$emit('toggle', this.nav.hover);
        }
      },
      beforeDestroy: function beforeDestroy() {
        this.nav.timer && window.clearTimeout(this.nav.timer);
      },
      methods: {
        handleMouseOver: function handleMouseOver(e) {
          var _this = this;
          if (!this.nav.click) {
            this.nav.enterTimer = setTimeout(function () {
              _this.nav.hover = true;
              _this.nav.enterTimer && window.clearTimeout(_this.nav.enterTimer);
              _this.nav.timer && window.clearTimeout(_this.nav.timer);
              _this.$emit('hover', _this.nav.hover);
              _this.$emit('toggle', _this.nav.hover);
            }, this.hoverEnterDelay);
          }
        },
        handleMouseLeave: function handleMouseLeave(e) {
          var _this2 = this;
          if (!this.nav.click) {
            this.nav.enterTimer && window.clearTimeout(this.nav.enterTimer);
            this.nav.timer = setTimeout(function () {
              _this2.nav.hover = false;
              window.clearTimeout(_this2.nav.timer);
              _this2.$emit('leave', _this2.nav.hover);
              _this2.$emit('toggle', _this2.nav.hover);
            }, this.hoverLeaveDelay);
          }
        },
        handleClick: function handleClick() {
          this.nav.click = !this.nav.click;
          this.nav.hover = this.nav.click;
          this.$emit('toggle', this.nav.hover);
          this.$emit('toggle-click', this.nav.hover);
        }
      }
    };

    /* script */
    var __vue_script__$1 = script$1;
    /* template */

    var __vue_render__$1 = function __vue_render__() {
      var _vm = this;

      var _h = _vm.$createElement;

      var _c = _vm._self._c || _h;

      return _c('div', _vm._b({
        staticClass: "bk-navigation"
      }, 'div', _vm.$attrs, false), [_vm.navigationType === 'top-bottom' ? _c('div', {
        staticClass: "bk-navigation-header",
        style: {
          backgroundColor: _vm.headThemeColor,
          flexBasis: _vm.headHeight + 'px'
        }
      }, [_c('navigation-menu-title', {
        attrs: {
          "side-title": _vm.sideTitle
        }
      }, [_vm.$slots['side-header'] || _vm.$slots['side-icon'] ? [_vm.$slots['side-header'] ? _vm._t("side-header", [_vm.$slots['side-icon'] ? _c('template', {
        slot: "side-icon"
      }, [_vm._t("side-icon")], 2) : _vm._e()]) : _vm.$slots['side-icon'] ? _c('template', {
        slot: "side-icon"
      }, [_vm._t("side-icon")], 2) : _vm._e()] : _vm._e()], 2), _c('div', {
        staticClass: "header-right"
      }, [_vm._t("header")], 2)], 1) : _vm._e(), _c('div', {
        staticClass: "bk-navigation-wrapper"
      }, [_vm.needMenu ? _c('div', {
        staticClass: "navigation-nav",
        style: {
          width: !_vm.nav.click ? _vm.navWidth + "px" : _vm.hoverWidth + "px"
        }
      }, [_c('div', {
        staticClass: "nav-slider",
        style: {
          width: !_vm.nav.hover ? _vm.navWidth + "px" : _vm.hoverWidth + "px",
          backgroundColor: _vm.defaultThemeColor,
          borderRight: _vm.navigationType !== 'top-bottom' ? 'none' : '1px solid #DCDEE5'
        },
        on: {
          "&mouseenter": function mouseenter($event) {
            return _vm.handleMouseOver($event);
          },
          "&mouseleave": function mouseleave($event) {
            return _vm.handleMouseLeave($event);
          }
        }
      }, [_vm.navigationType !== 'top-bottom' ? [_c('navigation-menu-title', {
        style: {
          flexBasis: _vm.headHeight + 'px'
        },
        attrs: {
          "side-title": _vm.sideTitle
        }
      }, [_vm.$slots['side-header'] || _vm.$slots['side-icon'] ? [_vm.$slots['side-header'] ? _vm._t("side-header", [_vm.$slots['side-icon'] ? _c('template', {
        slot: "side-icon"
      }, [_vm._t("side-icon")], 2) : _vm._e()]) : _vm.$slots['side-icon'] ? _c('template', {
        slot: "side-icon"
      }, [_vm._t("side-icon")], 2) : _vm._e()] : _vm._e()], 2)] : _vm._e(), _c('div', {
        staticClass: "nav-slider-list",
        style: {
          height: "calc(100vh - " + (_vm.headHeight + 56) + "px)"
        }
      }, [_vm._t("menu")], 2), _c('div', {
        staticClass: "nav-slider-footer"
      }, [_c('div', {
        staticClass: "footer-icon",
        class: {
          'is-left': _vm.navigationType !== 'top-bottom'
        },
        on: {
          "click": function click($event) {
            $event.stopPropagation();
            return _vm.handleClick($event);
          }
        }
      }, [_c('span', {
        staticClass: "footer-icon-svg",
        style: {
          transform: _vm.nav.click ? 'rotate(180deg)' : 'rotate(0deg)'
        }
      }, [_c('svg', {
        staticStyle: {
          "width": "1em",
          "height": "1em",
          "vertical-align": "middle",
          "fill": "currentColor",
          "overflow": "hidden"
        },
        attrs: {
          "viewBox": "0 0 64 64",
          "version": "1.1",
          "xmlns": "http://www.w3.org/2000/svg"
        }
      }, [_c('path', {
        attrs: {
          "d": "M31,13.5l-4.77,4.77,9.54,9.55H4v6.36H35.76l-9.53,9.55L31,48.5,48.47,31Zm22.65,0v35H60v-35Z"
        }
      })])])])])], 2)]) : _vm._e(), _c('div', {
        staticClass: "navigation-container",
        style: {
          maxWidth: _vm.needMenu ? 'calc(100vw - 60px)' : '100vw'
        }
      }, [_vm.navigationType !== 'top-bottom' ? [_c('div', {
        staticClass: "container-header",
        style: {
          flexBasis: _vm.headHeight + 'px'
        }
      }, [_vm._t("header", [_c('div', {
        staticClass: "container-header-title"
      }, [_vm._v(_vm._s(_vm.headerTitle))]), _c('div', {
        staticClass: "container-header-sets"
      }, [_vm._t("header-set")], 2)])], 2)] : _vm._e(), _c('div', {
        staticClass: "container-content",
        style: {
          maxHeight: "calc(100vh - " + _vm.headHeight + "px)"
        }
      }, [_vm._t("default"), _c('div', {
        staticClass: "container-footer"
      }, [_vm._t("footer")], 2)], 2)], 2)])]);
    };

    var __vue_staticRenderFns__$1 = [];
    /* style */

    var __vue_inject_styles__$1 = undefined;
    /* scoped */

    var __vue_scope_id__$1 = undefined;
    /* module identifier */

    var __vue_module_identifier__$1 = undefined;
    /* functional template */

    var __vue_is_functional_template__$1 = false;
    /* style inject */

    /* style inject SSR */

    /* style inject shadow dom */

    var __vue_component__$1 = /*#__PURE__*/normalizeComponent_1({
      render: __vue_render__$1,
      staticRenderFns: __vue_staticRenderFns__$1
    }, __vue_inject_styles__$1, __vue_script__$1, __vue_scope_id__$1, __vue_is_functional_template__$1, __vue_module_identifier__$1, false, undefined, undefined, undefined);

    var _defined = function (it) {
      if (it == undefined) throw TypeError("Can't call method on  " + it);
      return it;
    };

    var _toObject = function (it) {
      return Object(_defined(it));
    };

    var hasOwnProperty = {}.hasOwnProperty;
    var _has = function (it, key) {
      return hasOwnProperty.call(it, key);
    };

    var toString = {}.toString;
    var _cof = function (it) {
      return toString.call(it).slice(8, -1);
    };

    var _iobject = Object('z').propertyIsEnumerable(0) ? Object : function (it) {
      return _cof(it) == 'String' ? it.split('') : Object(it);
    };

    var _toIobject = function (it) {
      return _iobject(_defined(it));
    };

    var ceil = Math.ceil;
    var floor = Math.floor;
    var _toInteger = function (it) {
      return isNaN(it = +it) ? 0 : (it > 0 ? floor : ceil)(it);
    };

    var min = Math.min;
    var _toLength = function (it) {
      return it > 0 ? min(_toInteger(it), 0x1fffffffffffff) : 0;
    };

    var max = Math.max;
    var min$1 = Math.min;
    var _toAbsoluteIndex = function (index, length) {
      index = _toInteger(index);
      return index < 0 ? max(index + length, 0) : min$1(index, length);
    };

    var _arrayIncludes = function (IS_INCLUDES) {
      return function ($this, el, fromIndex) {
        var O = _toIobject($this);
        var length = _toLength(O.length);
        var index = _toAbsoluteIndex(fromIndex, length);
        var value;
        if (IS_INCLUDES && el != el) while (length > index) {
          value = O[index++];
          if (value != value) return true;
        } else for (;length > index; index++) if (IS_INCLUDES || index in O) {
          if (O[index] === el) return IS_INCLUDES || index || 0;
        } return !IS_INCLUDES && -1;
      };
    };

    function createCommonjsModule(fn, module) {
    	return module = { exports: {} }, fn(module, module.exports), module.exports;
    }

    var _core = createCommonjsModule(function (module) {
    var core = module.exports = { version: '2.6.12' };
    if (typeof __e == 'number') __e = core;
    });
    var _core_1 = _core.version;

    var _global = createCommonjsModule(function (module) {
    var global = module.exports = typeof window != 'undefined' && window.Math == Math
      ? window : typeof self != 'undefined' && self.Math == Math ? self
      : Function('return this')();
    if (typeof __g == 'number') __g = global;
    });

    var _shared = createCommonjsModule(function (module) {
    var SHARED = '__core-js_shared__';
    var store = _global[SHARED] || (_global[SHARED] = {});
    (module.exports = function (key, value) {
      return store[key] || (store[key] = value !== undefined ? value : {});
    })('versions', []).push({
      version: _core.version,
      mode:  'pure' ,
      copyright: '© 2020 Denis Pushkarev (zloirock.ru)'
    });
    });

    var id = 0;
    var px = Math.random();
    var _uid = function (key) {
      return 'Symbol('.concat(key === undefined ? '' : key, ')_', (++id + px).toString(36));
    };

    var shared = _shared('keys');
    var _sharedKey = function (key) {
      return shared[key] || (shared[key] = _uid(key));
    };

    var arrayIndexOf = _arrayIncludes(false);
    var IE_PROTO = _sharedKey('IE_PROTO');
    var _objectKeysInternal = function (object, names) {
      var O = _toIobject(object);
      var i = 0;
      var result = [];
      var key;
      for (key in O) if (key != IE_PROTO) _has(O, key) && result.push(key);
      while (names.length > i) if (_has(O, key = names[i++])) {
        ~arrayIndexOf(result, key) || result.push(key);
      }
      return result;
    };

    var _enumBugKeys = (
      'constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf'
    ).split(',');

    var _objectKeys = Object.keys || function keys(O) {
      return _objectKeysInternal(O, _enumBugKeys);
    };

    var _aFunction = function (it) {
      if (typeof it != 'function') throw TypeError(it + ' is not a function!');
      return it;
    };

    var _ctx = function (fn, that, length) {
      _aFunction(fn);
      if (that === undefined) return fn;
      switch (length) {
        case 1: return function (a) {
          return fn.call(that, a);
        };
        case 2: return function (a, b) {
          return fn.call(that, a, b);
        };
        case 3: return function (a, b, c) {
          return fn.call(that, a, b, c);
        };
      }
      return function () {
        return fn.apply(that, arguments);
      };
    };

    var _isObject = function (it) {
      return typeof it === 'object' ? it !== null : typeof it === 'function';
    };

    var _anObject = function (it) {
      if (!_isObject(it)) throw TypeError(it + ' is not an object!');
      return it;
    };

    var _fails = function (exec) {
      try {
        return !!exec();
      } catch (e) {
        return true;
      }
    };

    var _descriptors = !_fails(function () {
      return Object.defineProperty({}, 'a', { get: function () { return 7; } }).a != 7;
    });

    var document = _global.document;
    var is = _isObject(document) && _isObject(document.createElement);
    var _domCreate = function (it) {
      return is ? document.createElement(it) : {};
    };

    var _ie8DomDefine = !_descriptors && !_fails(function () {
      return Object.defineProperty(_domCreate('div'), 'a', { get: function () { return 7; } }).a != 7;
    });

    var _toPrimitive = function (it, S) {
      if (!_isObject(it)) return it;
      var fn, val;
      if (S && typeof (fn = it.toString) == 'function' && !_isObject(val = fn.call(it))) return val;
      if (typeof (fn = it.valueOf) == 'function' && !_isObject(val = fn.call(it))) return val;
      if (!S && typeof (fn = it.toString) == 'function' && !_isObject(val = fn.call(it))) return val;
      throw TypeError("Can't convert object to primitive value");
    };

    var dP = Object.defineProperty;
    var f = _descriptors ? Object.defineProperty : function defineProperty(O, P, Attributes) {
      _anObject(O);
      P = _toPrimitive(P, true);
      _anObject(Attributes);
      if (_ie8DomDefine) try {
        return dP(O, P, Attributes);
      } catch (e) {  }
      if ('get' in Attributes || 'set' in Attributes) throw TypeError('Accessors not supported!');
      if ('value' in Attributes) O[P] = Attributes.value;
      return O;
    };
    var _objectDp = {
    	f: f
    };

    var _propertyDesc = function (bitmap, value) {
      return {
        enumerable: !(bitmap & 1),
        configurable: !(bitmap & 2),
        writable: !(bitmap & 4),
        value: value
      };
    };

    var _hide = _descriptors ? function (object, key, value) {
      return _objectDp.f(object, key, _propertyDesc(1, value));
    } : function (object, key, value) {
      object[key] = value;
      return object;
    };

    var PROTOTYPE = 'prototype';
    var $export = function (type, name, source) {
      var IS_FORCED = type & $export.F;
      var IS_GLOBAL = type & $export.G;
      var IS_STATIC = type & $export.S;
      var IS_PROTO = type & $export.P;
      var IS_BIND = type & $export.B;
      var IS_WRAP = type & $export.W;
      var exports = IS_GLOBAL ? _core : _core[name] || (_core[name] = {});
      var expProto = exports[PROTOTYPE];
      var target = IS_GLOBAL ? _global : IS_STATIC ? _global[name] : (_global[name] || {})[PROTOTYPE];
      var key, own, out;
      if (IS_GLOBAL) source = name;
      for (key in source) {
        own = !IS_FORCED && target && target[key] !== undefined;
        if (own && _has(exports, key)) continue;
        out = own ? target[key] : source[key];
        exports[key] = IS_GLOBAL && typeof target[key] != 'function' ? source[key]
        : IS_BIND && own ? _ctx(out, _global)
        : IS_WRAP && target[key] == out ? (function (C) {
          var F = function (a, b, c) {
            if (this instanceof C) {
              switch (arguments.length) {
                case 0: return new C();
                case 1: return new C(a);
                case 2: return new C(a, b);
              } return new C(a, b, c);
            } return C.apply(this, arguments);
          };
          F[PROTOTYPE] = C[PROTOTYPE];
          return F;
        })(out) : IS_PROTO && typeof out == 'function' ? _ctx(Function.call, out) : out;
        if (IS_PROTO) {
          (exports.virtual || (exports.virtual = {}))[key] = out;
          if (type & $export.R && expProto && !expProto[key]) _hide(expProto, key, out);
        }
      }
    };
    $export.F = 1;
    $export.G = 2;
    $export.S = 4;
    $export.P = 8;
    $export.B = 16;
    $export.W = 32;
    $export.U = 64;
    $export.R = 128;
    var _export = $export;

    var _objectSap = function (KEY, exec) {
      var fn = (_core.Object || {})[KEY] || Object[KEY];
      var exp = {};
      exp[KEY] = exec(fn);
      _export(_export.S + _export.F * _fails(function () { fn(1); }), 'Object', exp);
    };

    _objectSap('keys', function () {
      return function keys(it) {
        return _objectKeys(_toObject(it));
      };
    });

    var keys = _core.Object.keys;

    var keys$1 = keys;

    function setInstaller (component, afterInstall) {
      component.install = function (Vue) {
        var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
        var props = component.props || {};
        keys$1(options).forEach(function (key) {
          if (props.hasOwnProperty(key)) {
            if (typeof props[key] === 'function' || props[key] instanceof Array) {
              props[key] = {
                type: props[key],
                default: options[key]
              };
            } else {
              props[key].default = options[key];
            }
          }
        });
        component.name = options.namespace ? component.name.replace('bk', options.namespace) : component.name;
        Vue.component(component.name, component);
        typeof afterInstall === 'function' && afterInstall(Vue, options);
      };
    }

    setInstaller(__vue_component__$1);

    exports.default = __vue_component__$1;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
